### ISSUE TITLE
e621 tag searching no result. Error parsing XML file.

### ISSUE THREAD

**Issue by JohnSteel163** [Bionus/imgbrd-grabber#733]
Quick personal comment, I love this app, it's almost perfect...
Most websites I try to download from work perfect but some of them I've been having problems, one of them is e621.
Here's the log with an xml error and wrong URL for e621 posts:

```
[12:56:17.471] Loading results...
[12:56:17.471] Loading page http://e621.net/posts.xml?limit=20&amp;page=1&amp;tags=dragon
[12:56:17.592] Error parsing XML file: unexpected character (37 - 127).
[12:56:17.592] Loading using Xml failed. Retry using Json.
[12:56:17.622] Error parsing JSON file: "

Attention Required! | CloudFlare

<meta http-equiv="X-UA-Compatible" "
[12:56:17.622] Loading using Json failed. Retry using Html.
[12:56:17.672] Receiving page http://e621.net/posts?limit=20&amp;page=1&amp;tags=dragon
```

In Grabber the search shows up as "No result". Tried different tags several times just in case, too.

I've tried correcting the URL problem myself to see if it'd work but no luck. The error "Error parsing XML file: unexpected character (37 - 127)." is still there even when I modified model.xml and I obviously don't understand what it is or what I'm doing as I really only know basics. I've backed up model.xml and returned the original after my attempts.

And as far as I know and remember the correct urls on e621 currently would be (not including the other items from the xml scripts of grabber or API):

```
/post/index?limit={limit}&amp;/{page}/{tags}
OR EVEN
/post?limit={limit}&amp;page={page}&amp;tags={tags}
```

As opposed to what was shown in some lines of the log
more specifically: **/post** vs **/posts**

Using latest version of Grabber on Windows XP SP3

Sorry if it's too long, just wanted to make sure I included all the details, and if this is already known, I've looked everywhere and found nothing like this.

**Comment by Bionus** [comment]
Have you tried manually visiting the url? (in this case http://e621.net/posts.xml?limit=20&amp;page=1&amp;tags=dragon)

Seems like e621 is using Cloudflare and that it was requiring your attention: `Attention Required! | CloudFlare`. Usually that's for a captcha page, for example if your browser/IP have been marked as suspicious either for making too many requests or because of where you live.

**Comment by JohnSteel163** [comment]
I decided to look into the issue again and do more indepth analysis and all I can find by myself, besides the fact that the URLs are wrong in the model.xml file, is that you **Really need to update Grabber to work with e621.net's current setup (ouroboro)** without depending on danbooru 2.0's settings, to put it simply. I'll supply you with my findings if they are of any use but I am not able to do anything else by myself without more info from Grabber, so **please read last paragraph for URLs that should be fixed besides the search URLs**, everything else is simply me stating that it really isn't a problem with my computer, setup, connection or cloudflare. 

`/posts.xml?`... and `/post.xml?` don't work on my browser and gives me 404, only `/post?...` shows posts in browser, meaning two things: That cloudflare alert in the log was simply a spam block measure caused by the error and not the other way around, it really didn't ask me for a captcha when I browsed the website. Source 1 option in source > e621.net > options can be changed to RSS and the XML and JSON errors can be completely ignored since it won't be using XML or JSON source, and this also fixes cloudflare requiring attention, but still doesn't fix the main issue.

When I fixed the search URLs in the model.xml file, the log says it's retrieving pictures from e621 but it does not show any results and says "no result", most likely because not only the search URLs are (were) incorrect but also the file URLs, sample URLs and preview URLs, all are incorrect, so it looks like this (and again, no results):

```
[20:25:29.474] Loading results...
[20:25:29.474] Loading page http://e621.net/post?limit=20&amp;page=1&amp;tags=
[20:25:30.565] Receiving page https://e621.net/post?limit=20&amp;page=1&amp;tags=
```

So I've looked into getting more info on it and I did, and also tried to on how grabber works but failed so far. All I can conclude so far is that the file URLs in the model.xml SHOULD BE:
**`https://static1.e621.net/data/{2 first numbers of the md5}/{2 following numbers of the md5}/{md5}.{ext}`**
As shown in the following example: `/data/ab/12/ab123456cdefghxxxxxxx...wxyz.ext`
Or so this is the case for ALL of the pictures I've seen the URLs of. I don't know how to add that request in the model.xml. Maybe md5maxlegth=2 then something similar for the next number after the slash?
The same works for preview and sample:
`/data/preview/12/34/md5.ext` and `/data/sample/12/34/md5.ext`
There's also the question whether the URLs to the files (all examples above) starting as `https://static1.e621.net` make any difference, which it probably does...

**Comment by Bionus** [comment]
I'll take a look as soon as I have some free time as e621 is an important source for Grabber users, and if need be will separate it from the Danbooru source and give it its own.
It seems that they are making more and more updates these days. The posts/post issue wasn't a problem before as both used to return results (or had some redirection) so I didn't really look into it before.

As for the XML version, that would mean that they either disabled it or moved it behind a login wall (have you tried authenticating for using it?). I'll check their API doc if they have it up to date.
For the images URLs, those are not really set in stone, they should be directly got from the results listing page or inferred from the thumbnails' urls.

In all cases, thanks for already clearing a lot of the problems beforehand, that'll speed up my fix considerably.

**Comment by JohnSteel163** [comment]
The API doc says they still have .xml and .json enabled so I'll try that once everything is in the right place but right now that won't work either. Thank you and you're welcome.

If I can help with anything else, I have a lot of free time this week, I would just need some info on the script, commands, functions, requests, and I haven't found much about it. I could write it all myself if only I did.

**Comment by Bionus** [comment]
Actually, could it be that you've added e621 as a Danbooru 2.0 source while it's a Danbooru 1.0 one?

**Comment by JohnSteel163** [comment]
It was added as Danbooru 2.0 as default and Danbooru 1.0 doesn't work either.

**Comment by Bionus** [comment]
Indeed, e621 is adding some stuff on their interface, and have a few new urls. However, they didn't remove the old Danbooru 1 urls.

I just tried and downloaded with no issues a bunch of images (`solo rating:safe`), even with detailed tag information: `%species%/%artist%/%md5%.%ext%`, using it as a Danbooru 1 source (XML > JSON > Regex). The urls returned for that download were:

```
http://e621.net/post/index?limit=20&page=1&tags=solo rating%3Asafe
http://e621.net/post/index.xml?limit=20&page=1&tags=solo rating%3Asafe
http://e621.net/post/index.json?limit=20&page=1&tags=solo rating%3Asafe
```

All of those match the Danbooru 1 url patterns and don't even require authentication to work.

Do you have more details about what happens?
I really cannot reproduce your issue.

**Comment by JohnSteel163** [comment]
Before 5.1.0 what I got from using Danbooru 1 the url pattern was out of the correct order and had some unecessary items in it but apparently that isn't a problem anymore, I've tested now that I got some free time again and it's working. And again, e621.net was added as danbooru 2 as default when I first installed the program, since it does look like it'd work having all the correct urls. Should this be closed now?
